src/platform_storage_api/admission_controller/api.py

```diff
@@ -0,0 +1,219 @@
+
+        pod = payload["request"]["object"]
+        kind = pod.get("kind")
+
+        if kind != "Pod":
+            # not a pod creation request. early-exit
+            return admission_review.to_response()
+
+        metadata = pod.get("metadata", {})
+        labels = metadata.get("labels")
+
+        # let's check if this is a new platform storage POD
+        if (
+            labels.get("app") == LABEL_PLATFORM_STORAGE_POD and
+            labels.get("service") == LABEL_PLATFORM_STORAGE_POD
+        ):
+            return await self._handle_new_platform_storage_pod(
+                pod=pod,
+                admission_review=admission_review,
+            )
+
+        annotations = metadata.get("annotations", {})
+        raw_injection_spec = annotations.get(ANNOTATION_APOLO_INJECT_STORAGE)
+
+        if not raw_injection_spec:
+            # a pod does not request storage. we can do early-exit here
+            logger.info("POD won't be mutated because doesnt define proper annotations")
+            return admission_review.to_response()
+
+        return await self._handle_injection(
+            pod=pod,
+            raw_injection_spec=raw_injection_spec,
+            admission_review=admission_review,
+        )
+
+    async def _handle_new_platform_storage_pod(
+        self,
+        pod: dict[str, Any],
+        admission_review: AdmissionReviewResponse,
+    ) -> web.Response:
+        await self._volume_resolver.refresh_internal_state(pod=pod)
+        return admission_review.to_response()
+
+    async def _handle_injection(
+        self,
+        pod: dict[str, Any],
+        raw_injection_spec: str,
+        admission_review: AdmissionReviewResponse,
+    ) -> web.Response:
+        pod_spec = pod["spec"]
+        containers = pod_spec.get("containers") or []
+
+        if not containers:
+            logger.info("POD won't be mutated because doesnt define containers")
+            # pod does not define any containers. we can exit
+            return admission_review.to_response()
+
+        logger.info("Going to inject volumes")
+        try:
+            injection_spec = json.loads(raw_injection_spec)
+        except ValueError:
+            logger.info("Invalid injection spec. Denying the request")
+            admission_review.allowed = False
+            return admission_review.to_response()
+
+        # let's ensure POD has volumes
+        if "volumes" not in pod_spec:
+            admission_review.add_patch(
+                path="/spec/volumes",
+                value=[]
+            )
+
+        # and ensure that each container has a volume mounts
+        for idx, container in enumerate(containers):
+            if "volumeMounts" not in container:
+                admission_review.add_patch(
+                    path=f"/spec/containers/{idx}/volumeMounts",
+                    value=[]
+                )
+
+        for mount_path, storage_path in injection_spec.items():
```

Comment:
We should also support volume mount modes (readonly, readwrite). We support them in cli

Reply:
added that